src/libs/Middleware/Reauthentication.js

```diff
@@ -0,0 +1,95 @@
+import lodashGet from 'lodash/get';
+import CONST from '../../CONST';
+import * as NetworkStore from '../Network/NetworkStore';
+import * as MainQueue from '../Network/MainQueue';
+import * as Authentication from '../Authentication';
+import * as PersistedRequests from '../actions/PersistedRequests';
+import * as Request from '../Request';
+import Log from '../Log';
+
+/**
+ * Reauthentication middleware
+ *
+ * @param {Promise} response
+ * @param {Object} request
+ * @param {Boolean} isFromSequentialQueue
+ * @returns {Promise}
+ */
+function Reauthentication(response, request, isFromSequentialQueue) {
+    return response
+        .then((data) => {
+            if (NetworkStore.isOffline()) {
+                // If we are offline and somehow handling this response we do not want to reauthenticate
+                throw new Error('Unable to reauthenticate because we are offline');
+            }
+
+            if (data.jsonCode === CONST.JSON_CODE.NOT_AUTHENTICATED) {
+                const credentials = NetworkStore.getCredentials();
+
+                // Credentials haven't been initialized. We will not be able to re-authenticate with the API.
+                const unableToReauthenticate = (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword);
+                if (unableToReauthenticate) {
+                    if (isFromSequentialQueue) {
+                        throw new Error('Missing credentials required for authentication');
+                    }
+
+                    MainQueue.replay(request);
```

Comment:
Yeah, this is a really great find and I think points to a deeper problem. 

The comment suggests that "Credentials haven't been initialized". So we are putting this request back into the queue because we are "waiting for the credentials". Which begs these questions: 

- Why we are waiting for them at all ?
- Why we are not waiting for them in the sequential queue ?

Gonna think about this for a second. My gut says no requests should happen at all until we can guarantee the credentials either exist or don't exist. Then we can just throw here and the user should probably also get signed out.

Reply:
Looked into this a bit. The changes required here are a bit involved but I think worth doing...

We have some logic here:

https://github.com/Expensify/App/blob/58a139ab96a6189a6e0dcbbd003af18e7d5194dc/src/libs/Network/MainQueue.js#L23-L27

and it needs a better design since both the "main queue" and "sequential queue" should only make requests when we are "ready" to do so.

Reply:
Going to create a follow up issue for this as it's slightly tricky to untangle and I don't think it needs to be a blocker.

Reply:
TL;DR - no, credentials are not automatically regenerated and this code appears related to a race condition. the queue we have for these requests now isn't "waiting for the credentials" so it's not clear if this is a problem or not, but should be cleaned up.

Reply:
Created an issue here: https://github.com/Expensify/Expensify/issues/208303

Reply:
👍